hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ServicesState.java

```diff
@@ -610,6 +627,73 @@ boolean shouldMigrateNfts() {
     }
 
+    private static void migrateVirtualMapsToMerkleDb(final ServicesState state) throws InterruptedException {
+        final VirtualMapFactory vmFactory = new VirtualMapFactory();
+
+        // virtualized blobs
+        final VirtualMap<VirtualBlobKey, VirtualBlobValue> storageMap =
+                state.getChild(StateChildIndices.STORAGE);
+        if (jasperDbBacked(storageMap)) {
+            VirtualMap<VirtualBlobKey, VirtualBlobValue> merkleDbBackedMap =
+                    vmFactory.newVirtualizedBlobs();
+            merkleDbBackedMap = migrateVirtualMap(storageMap, merkleDbBackedMap);
+            state.setChild(StateChildIndices.STORAGE, merkleDbBackedMap);
+        }
+
+        // virtualized iterable storage
+        final VirtualMap<ContractKey, IterableContractValue> contractStorageMap =
+                state.getChild(StateChildIndices.CONTRACT_STORAGE);
+        if (jasperDbBacked(contractStorageMap)) {
+            VirtualMap<ContractKey, IterableContractValue> merkleDbBackedMap =
+                    vmFactory.newVirtualizedIterableStorage();
+            merkleDbBackedMap = migrateVirtualMap(contractStorageMap, merkleDbBackedMap);
+            state.setChild(StateChildIndices.CONTRACT_STORAGE, merkleDbBackedMap);
+        }
+
+        // virtualized unique token storage, if enabled
+        if (state.enabledVirtualNft) {
+            final VirtualMap<UniqueTokenKey, UniqueTokenValue> uniqueTokensMap =
+                    state.getChild(StateChildIndices.UNIQUE_TOKENS);
+            if (jasperDbBacked(uniqueTokensMap)) {
+                VirtualMap<UniqueTokenKey, UniqueTokenValue> merkleDbBackedMap =
+                        vmFactory.newVirtualizedUniqueTokenStorage();
+                merkleDbBackedMap = migrateVirtualMap(uniqueTokensMap, merkleDbBackedMap);
+                state.setChild(StateChildIndices.UNIQUE_TOKENS, merkleDbBackedMap);
+            }
+        }
+    }
+
+    private static boolean jasperDbBacked(final VirtualMap<?, ?> map) {
+        final VirtualRootNode<?, ?> virtualRootNode = map.getRight();
+        return virtualRootNode.getDataSource() instanceof VirtualDataSourceJasperDB;
+    }
+
+    private static <K extends VirtualKey<? super K>, V extends VirtualValue>
+            VirtualMap<K, V> migrateVirtualMap(final VirtualMap<K, V> source, final VirtualMap<K, V> target
+    ) throws InterruptedException {
+        final int copyTargetMapEveryPuts = 10_000;
+        final AtomicInteger count = new AtomicInteger(copyTargetMapEveryPuts);
+        final AtomicReference<VirtualMap<K, V>> targetMapRef = new AtomicReference<>(target);
+        VirtualMapMigration.extractVirtualMapData(
+                AdHocThreadManager.getStaticThreadManager(),
+                source,
+                kvPair -> {
+                    final K key = kvPair.getKey();
+                    final V value = kvPair.getValue();
+                    log.info("Migrating " + key + " to " + value);
```

Comment:
There are hundreds of millions of k/v pairs in some perfnet states, so we definitely don't want to log each migration.

But we might wrap this `extractVirtualMapData()` call [with `withLoggedDuration()`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/utils/MiscUtils.java#L868) to get a record of how long it takes.

Reply:
Yes, I noticed this line produces tons of output. It's just for debugging purpose for myself, so I'll remove it before merging.